### PR TITLE
CPR-310 Add read / write lock 

### PR DESCRIPTION
**Pull request by H-Iwanejko**
Initial attempt to fix, TODO: add a wrapper around the PersonService so it handles future calls.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/message/processor/CourtCaseEventsProcessor.kt [hunk 1]
```diff
@@ -62,29 +62,28 @@ class CourtCaseEventsProcessor(
...
       }
     } catch (e: Exception) {
       when (e) {
         is CannotAcquireLockException, is JpaSystemException -> {
           log.warn("Expected error when processing $e.message")
-          telemetryService.trackEvent(
-            HMCTS_EXACT_MATCH,
-            mapOf("PNC" to person.otherIdentifiers?.pncIdentifier.toString(), "Exception" to e.message),
-          )
         }
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/message/processor/CourtCaseEventsProcessor.kt:82]: we should reprocess on all these exceptions

File: src/test/kotlin/uk/gov/justice/digital/hmpps/personrecord/message/listeners/CourtCaseEventsListenerIntTest.kt [hunk 2]
```diff
@@ -82,6 +83,11 @@ class CourtCaseEventsListenerIntTest : IntegrationTestBase() {
       CPR_RECORD_CREATED,
       mapOf("SourceSystem" to "HMCTS"),
     )
+    checkTelemetry(
+      CPR_RECORD_UPDATED,
+      mapOf("SourceSystem" to "HMCTS"),
+      times(99),
+    )
```
Comment (markreesmoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/personrecord/message/listeners/CourtCaseEventsListenerIntTest.kt:90]: ❤️

**Review by grnewbould — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/ReadWriteLockService.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import org.springframework.transaction.annotation.Transactional
+import java.util.concurrent.locks.ReentrantReadWriteLock
+
+@Service
+class ReadWriteLockService {
+
+  private val lock = ReentrantReadWriteLock()
+
+  @Transactional(isolation = Isolation.READ_COMMITTED, readOnly = true)
+  fun <T> withReadLock(block: () -> T?): T? {
```
[the remark this answers is not in the document]
Comment (H-Iwanejko) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/ReadWriteLockService.kt:11]: We are going to need this later when retrieving candidates

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/PersonService.kt [hunk 4]
```diff
@@ -13,16 +11,18 @@ import uk.gov.justice.digital.hmpps.personrecord.service.type.TelemetryEventType
...
-    val existingPersonEntity: PersonEntity? = callback()
-    when {
-      (existingPersonEntity == null) -> handlePersonCreation(person)
-      else -> handlePersonUpdate(person, existingPersonEntity)
+    readWriteLockService.withWriteLock {
+      val existingPersonEntity: PersonEntity? = callback()
+      when {
+        (existingPersonEntity == null) -> handlePersonCreation(person)
+        else -> handlePersonUpdate(person, existingPersonEntity)
+      }
```
[the remark this answers is not in the document]
Comment (H-Iwanejko) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/PersonService.kt:25]: Needs to be a write lock here as we retrieve and make changes to that entity. I have tried with Read + a write lock but still causes a small concurrency issue. But the wrapped write lock prevents that.

**Pull request merged**